### PR TITLE
feat: add metrics module aligned with unification

### LINKED ISSUE #4178: [DOCS] Create and end2end `examples` directory with `.ipynb` files that align with practical guid usage.
## Which page or section is this issue related to?
The idea is to have runnable end2end examples of basic real-life examples for the different `task-templates`, where we will initially focus on the most common task `text-classification`.

Things to keep in mind:
- they need to be executable individually but may require previous examples to be run by referencing them
- they need to be aligned with the [practical guides](https://docs.argilla.io/en/latest/practical_guides/practical_guides.html) to ensure maintainability
- it might be nice to add an [optional telemetry module](https://haystack.deepset.ai/tutorials/14_query_classifier) to see the number of usages for tutorials.
- ideally, an updated version of these tutorial datasets should be available on the Hugging Face hub as imports

image

```[tasklist]
### Tasks
- [ ] https://github.com/argilla-io/argilla/issues/4189
- [ ] https://github.com/argilla-io/argilla/issues/4198
- [ ] https://github.com/argilla-io/argilla/issues/4179
- [ ] https://github.com/argilla-io/argilla/issues/4183
- [ ] https://github.com/argilla-io/argilla/issues/4199
- [ ] https://github.com/argilla-io/argilla/issues/4181
- [ ] https://github.com/argilla-io/argilla/issues/4182
- [ ] https://github.com/argilla-io/argilla/issues/4185
- [ ] https://github.com/argilla-io/argilla/issues/4184
- [ ] https://github.com/argilla-io/argilla/issues/4186
- [ ] https://github.com/argilla-io/argilla/issues/4187
- [ ] https://github.com/argilla-io/argilla/issues/4220
- [ ] https://github.com/argilla-io/argilla/issues/4280
```

### LINKED ISSUE #4263: [REFACTOR] align structure and usage of `unification` with `metrics`
**Is your feature request related to a problem? Please describe.**
We aim to introduce the https://github.com/argilla-io/argilla/pull/4175, but ideally this should be aligned as much as possible with our existing `unification`

**Describe the solution you'd like**
N.A.

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by plaguss**
# Description

This PR aligns the metric modules for a dataset with the unification strategies.
It includes a new `UnifiedAnnotationMetric` class to compute the `AnnotatorMetric` for a *unified* dataset.

Should be merged after #4175.

Closes #4263, #4034

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [x] New feature (non-breaking change which adds functionality)
- [x] Refactor (change restructuring the codebase without changing functionality)
- [ ] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [x] `tests/integration/client/feedbac/metrics/test_annotator_metrics.py`

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

**Comment by plaguss** [comment]
Hi @davidberenstein1957, this PR starts from the branch of the other one. I've added methods to compute the metrics directly from the `FeedbackDataset`. If you think it works like that, I'll take a look to see in which of the parent classes/mixins fits better.

File: docs/_source/practical_guides/collect_responses.md [hunk 1]
```diff
@@ -120,6 +120,83 @@ Once you have unified your responses, you will have a dataset that's ready for [
...
+from argilla.client.feedback.metrics import AgreementMetric
+
+feedback_dataset = rg.FeedbackDataset.from_argilla("...", workspace="...")
+metric = AgreementMetric(dataset=feedback_dataset, question_name="question_name")
+agreement_metrics = metric.compute("alpha")
+# >>> agreement_metrics
+# [AgreementMetricResult(metric_name='alpha', result=0.0)]
+```
+
+Currently only [Krippendorf's alpha](https://en.wikipedia.org/wiki/Krippendorff%27s_alpha) is defined, for all the questions types except the `TextQuestion`. It applies to any number of annotators, with possible missing responses from them.
```
Comment (davidberenstein1957) [inline docs/_source/practical_guides/collect_responses.md]: "with possible missing responses from them." is not very clear to me

File: src/argilla/client/feedback/dataset/local/dataset.py [hunk 2]
```diff
@@ -627,3 +629,65 @@ def find_similar_records(
             UserWarning,
         )
         return []
+
+    def compute_annotator_metrics(
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/dataset/local/dataset.py]: This looks good. perhaps we should also rename `unify_responses` to `compute_unified_responses`? I think aligning naming.

**Review by davidberenstein1957 — COMMENTED** [review]
Hi @plaguss , this looks good already. If you want, feel free to also have a look at the structure of the code in general. I think it might be nice to structure the code in a similar way as the `metrics` module.

File: docs/_source/practical_guides/collect_responses.md [hunk 3]
```diff
@@ -189,15 +189,16 @@ We plan on adding more support for other metrics so feel free to reach out on ou
...
+| accuracy              | ✔️              | ✔️                  | ✔️              |                 |              |
+| precision             | ✔️              | ✔️                  | ✔️              |                 |              |
+| recalll               | ✔️              | ✔️                  | ✔️              |                 |              |
+| f1-score              | ✔️              | ✔️                  | ✔️              |                 |              |
+| confusion-matrix      | ✔️              | ✔️                  | ✔️              |                 |              |
 | pearson-r             | ✔️              |                    |                |                 |              |
 | spearman-r            |                |                    | ✔️              |                 |              |
 | gleu                  |                |                    |                |                 | ✔️            |
 | rouge                 |                |                    |                |                 | ✔️            |
+| ndcg-score            |                |                    |                | ✔️               | ✔️            |
```
Comment (davidberenstein1957) [inline docs/_source/practical_guides/collect_responses.md]: I believe this one does not apply for `TextQuestion`

Comment (plaguss) [reply docs/_source/practical_guides/collect_responses.md]: Ups, copy/paste error, will update it

**Comment by plaguss** [comment]
We need to merge the following PRs before this one:
- #4314 - To see the tests in the CI before.
- #4295 - To solve the tests failing with the ranking questions.

File: src/argilla/client/feedback/dataset/local/dataset.py [hunk 4]
```diff
@@ -418,15 +421,15 @@ def delete_metadata_properties(
         self._metadata_properties = list(metadata_properties_mapping.values())
         return deleted_metadata_properties if len(deleted_metadata_properties) > 1 else deleted_metadata_properties[0]
 
-    def unify_responses(
+    def compute_unified_responses(
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/dataset/local/dataset.py]: can you add a depracation warning for this?

File: src/argilla/client/feedback/dataset/local/dataset.py [hunk 5]
```diff
@@ -63,7 +64,9 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-class FeedbackDataset(ArgillaMixin, HuggingFaceDatasetMixin, FeedbackDatasetBase[FeedbackRecord], TaskTemplateMixin):
+class FeedbackDataset(
+    ArgillaMixin, HuggingFaceDatasetMixin, FeedbackDatasetBase[FeedbackRecord], TaskTemplateMixin, MetricsMixin
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/dataset/local/dataset.py]: can you also create a unification mixin?

File: src/argilla/client/feedback/dataset/local/dataset.py (hunk 4 above)
Comment (plaguss) [reply src/argilla/client/feedback/dataset/local/dataset.py]: Sorry, I forgot it. It's done, I opened a new issue to test the deprecation and add notes in the documentation for that: #4322

File: src/argilla/client/feedback/dataset/local/dataset.py (hunk 5 above)
Comment (plaguss) [reply src/argilla/client/feedback/dataset/local/dataset.py]: Done!

File: src/argilla/client/feedback/dataset/local/dataset.py (hunk 4 above)
Comment (plaguss) [reply src/argilla/client/feedback/dataset/local/dataset.py]: Done

File: docs/_source/practical_guides/collect_responses.md [hunk 6]
```diff
@@ -120,6 +120,124 @@ Once you have unified your responses, you will have a dataset that's ready for [
...
+| f1-score              | ✔️              | ✔️                  | ✔️              |                 |              |
+| confusion-matrix      | ✔️              | ✔️                  | ✔️              |                 |              |
+| pearson-r             | ✔️              |                    |                |                 |              |
+| spearman-r            |                |                    | ✔️              |                 |              |
+| gleu                  |                |                    |                |                 | ✔️            |
+| rouge                 |                |                    |                |                 | ✔️            |
+| ndcg-score            |                |                    |                | ✔️               |              |
+
+
+#### Unified annotation metrics
```
Comment (davidberenstein1957) [inline docs/_source/practical_guides/collect_responses.md]: I think we should use `UnifiedAnnotatorMetric` to align with `AnnotatorMetric`

File: docs/_source/reference/python/python_annotation_metrics.rst [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+Annotator Metrics
+-----------------
+
+.. automodule:: argilla.client.feedback.metrics.annotator_metrics
+   :members:
+
+.. autoclass:: argilla.client.feedback.metrics.annotator_metrics.AnnotatorMetric
+   :members: __init__, compute
+
+.. autoclass:: argilla.client.feedback.metrics.annotator_metrics.ResponsesMetric
```
Comment (davidberenstein1957) [inline docs/_source/reference/python/python_annotation_metrics.rst]: maybe we can also unify the usage of plural vs singular here? In some cases we use plural and in others singular. 
`Annotator` vs `Suggestions`

Comment (davidberenstein1957) [reply docs/_source/reference/python/python_annotation_metrics.rst]: Intuitively, I would go for all singular.

File: docs/_source/practical_guides/collect_responses.md [hunk 8]
```diff
@@ -120,6 +120,124 @@ Once you have unified your responses, you will have a dataset that's ready for [
...
+| recalll               | ✔️              | ✔️                  | ✔️              |                 |              |
+| f1-score              | ✔️              | ✔️                  | ✔️              |                 |              |
+| confusion-matrix      | ✔️              | ✔️                  | ✔️              |                 |              |
+| pearson-r             | ✔️              |                    |                |                 |              |
+| spearman-r            |                |                    | ✔️              |                 |              |
+| gleu                  |                |                    |                |                 | ✔️            |
+| rouge                 |                |                    |                |                 | ✔️            |
+| ndcg-score            |                |                    |                | ✔️               |              |
+
+
```
Comment (davidberenstein1957) [inline docs/_source/practical_guides/collect_responses.md:265]: Can you update the docs to include the Suggestion and ResponseMetric

Comment (plaguss) [reply docs/_source/practical_guides/collect_responses.md:265]: On it!

File: docs/_source/practical_guides/collect_responses.md (hunk 6 above)
Comment (plaguss) [reply docs/_source/practical_guides/collect_responses.md]: Done

File: docs/_source/reference/python/python_annotation_metrics.rst (hunk 7 above)
Comment (plaguss) [reply docs/_source/reference/python/python_annotation_metrics.rst]: Sure

File: docs/_source/practical_guides/collect_responses.md (hunk 8 above)
Comment (plaguss) [reply docs/_source/practical_guides/collect_responses.md:265]: Those will be the same for ResponsesMetric and SuggestionMetrics I guess.

File: src/argilla/client/feedback/dataset/local/dataset.py (hunk 5 above)
Comment (plaguss) [reply src/argilla/client/feedback/dataset/local/dataset.py]: 👍

**Comment by davidberenstein1957** [comment]
I have a problem with unifying responses from remote datasets while local ones work perfectly. I tried Ranking, Ranking, Label and Multilabel and all of them returned the error. I don't know if you have encountered this problem but here is the code to reproduce it.

```
ranking_ds = rg.FeedbackDataset.from_argilla("ranking_ds")
ranking_ds_local = ranking_ds.pull()

# RANKING - REMOTE - ERROR
ranking_ds.compute_unified_responses(question="preference", strategy="min")
for record in ranking_ds:
    print(record.responses)
    print(record.unified_responses)

# RANKING - LOCAL - WORKS FINE
ranking_ds_local.compute_unified_responses(question="preference", strategy="min") 
for record in ranking_ds_local:
    print(record.responses)
    print(record.unified_responses)

```

Also, here are the dummy test datasets that I worked with.

https://huggingface.co/datasets/kursathalat/multilabel_ds
https://huggingface.co/datasets/kursathalat/label_ds
https://huggingface.co/datasets/kursathalat/rating_ds
https://huggingface.co/datasets/kursathalat/ranking_ds

**Review by kursathalat — APPROVED** [review]

File: docs/_source/practical_guides/collect_responses.md [hunk 9]
```diff
@@ -203,8 +203,8 @@ We obtain a `dict` where the keys contain the `user_id` of a given annotator and
 Alternatively, we have the opportunity to compute the metrics directly from the dataset. Let’s use the following dataset for this, and compute the metrics for the suggestions:
 
 ```python
-suggestions_metrics = dataset.compute_model_metrics(question_name="label", metric_names=["accuracy", "precision", "recall", "f1-score"])
-suggestions_metrics['00000000-0000-0000-0000-000000000001']
+mmodel_metrics = dataset.compute_model_metrics(question_name="label", metric_names=["accuracy", "precision", "recall", "f1-score"])
+mmodel_metrics['00000000-0000-0000-0000-000000000001']
```
Comment (plaguss) [inline docs/_source/practical_guides/collect_responses.md]: typo in `mmodel_metrics`

**Review by plaguss — COMMENTED** [review]
nice job!! 🙌

**Pull request merged**